src/compiler/checker.ts

```diff
@@ -21331,6 +21331,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 tracing?.pop();
             }
 
+            // if (entry !== undefined) {
+            //     // If the previous entry and the result disagree, then something has gone wrong.
+            //     Debug.assert(!!(entry & RelationComparisonResult.Succeeded) === (result !== Ternary.False), "Cached relationship does not match recalculated result");
+            // }

```

Comment:
I have this commented out as I used it to debug things, but it feels like a valuable consistency check that could be left in.